### DISCUSSION TITLE
Problems setting sound card

### DISCUSSION THREAD

**Discussion by w0mbatian** [mcguirepr89/BirdNET-Pi#551] [Q&A]
Hi. Just getting a new setup going and this is my first post here.
I am using a Zoom H1 as a USB microphone which seems to work well with BirdNET-Pi.
It is recognised in alsamixer. However when I go back to alsamixer it is no longer set and the default is showing again.
I then found the box to set sound card in advanced settings. I replaced the word default with H1 which is what shows in alsamixer.
However saving this setting immediately stops analysis. If I change H1 back to default analysis starts again.
Am I doing something wrong?
Ian

**Comment by mcguirepr89** [comment]
Hi, @w0mbatian --

I'm not 100% clear on this part: 
>However when I go back to alsamixer it is no longer set and the default is showing again.

Did analysis stop, all of a sudden? Is that what that means? (I'm trying to figure out where BirdNET-Pi messed up so I can figure out how to help)

For setting the REC_CARD variable in Advanced Settings, "`Set Audio Card to 'default' to use PulseAudio (always recommended), or an ALSA recognized sound card device from the output of `aplay -L`.`", so you'll want to issue `aplay -L` (or, to be more semantically consistent, `arecord -L`) and use output from that, not the alsamixer.

From that output, try to use a `dsnoop:` option to allow the analysis and livestream to share the device -- otherwise, if your device driver won't create the `dsnoop:` device, you'll need to use the `hw:` device and disable the livestream in "Tools" > "Services" so that they don't fight over access to the hardware. (Or keep using `default` and let me know more about what you're seeing and we can look into adjusting settings or something along those lines)

My best!

**Comment by w0mbatian** [comment]
Analysis stops when I set H1 (as detected in alsamixer) in Audio Card in Advanced settings. Analysis starts again if I set Audio Card back to default.
In alsamixer the Zoom H1 is detected properly as H1 . If I choose H1 from the list of devices  it then appears in the top left corner. If I then exit alsamixer and then go back in the setting in the top left has reverted to default leaving me in doubt as to whether the correct sound device is selected or not.
So there are two related things I am asking about.
I'm afraid I don't understand most of the rest of what you said. I have minimal linux knowledge.
Ian

**Comment by w0mbatian** [comment]
I have just realised those commands are for the web terminal.
aplay -L gives a lot of output at then end of which is
usbstream:CARD=H1
H1
USB Stream Output

That is what I have been setting in Audio Card in advanced settings. Setting this causes analysis to immediately stop and it stays stpped until Audio Card is set back to "default".

arecord -L gives no output.
Ian

**Reply by mcguirepr89** [reply]
> If I choose H1 from the list of devices it then appears in the top left corner. If I then exit alsamixer and then go back in the setting in the top left has reverted to default leaving me in doubt as to whether the correct sound device is selected or not.
So there are two related things I am asking about.

I understand, now, and hope I can make things a bit more clear as it is all a bit esoteric (ALSA and PulseAudio aren't known for being the most straightforward things on Earth, that's for sure!).

`alsamixer` is only a mixer for the available ALSA devices, both Playback and Capture devices. When PulseAudio is installed, however, pulse naturally becomes the wrapper around all ALSA device and becomes your virtualized `default` device. This means utilities that ask ALSA which device to use actually get their response from Pulse in the form of its default (a pulse wrapper around whatever is set as the default playback and capture devices for alsa).

If you only have one USB mic plugged in, it will be the default for both ALSA and Pulse, and you won't need to do anything to change settings. If you want to set, for instance, the capture level for your _alsa_ device, you could use `alsamixer` for that, but you would probably then want to bypass PulseAudio altogether and access the ALSA capture device directly.

However . . .

>arecord -L gives no output.

I have NO idea how your `default` pulse audio device is working in the first place if ALSA detects no capture devices. 

I'm afraid this TOTALLY confounds my understanding of the relationship between ALSA and Pulse. 

It is my understanding that if `arecord -L` does not show your device as output, you won't be able to use that device.

Since, however, you say that the system works when the REC_CARD is set to its default (pulseaudio/'default'), I really just don't know at all **how**????

PS
The sound card name that shows up in `alsamixer` is for humans to read -- when setting sound card settings (anywhere), you need to identify it by its device name, which is provided in the output of `arecord -L`

**Comment by w0mbatian** [comment]
I have no idea why but this morning I do get output from arecord -L. Apologies for whatever stuff up I did.
I have attached a screen grab of most of the output.
![arecord-L output](https://user-images.githubusercontent.com/113158830/190933268-e976d9de-5199-46ba-a507-689ba1f16995.jpg)

**Reply by mcguirepr89** [reply]
That is a relief -- I was worried something fundamental had changed and that I was unaware.

From the output of `arecord -L` that you've provided, the _un_-indented lines are device names. So, for instance, your USB mic's `dnsoop:` device is `dsnoop:CARD=H1,DEV=0`

If you input the device name, I'm pretty confident everything will work, though I'm still unsure on the _why_ of it all.

**Comment by w0mbatian** [comment]
"If you input the device name, I'm pretty confident everything will work, though I'm still unsure on the why of it all."

Do you mean to put H1 in the advanced settings in the Audio Card box?
If so, that is when analysis stops.

**Reply by mcguirepr89** [reply]
You simply must read the entire comment PLEASE!

>From the output of `arecord -L` that you've provided, **the un-indented lines are device names**. So, for instance, your USB mic's `dnsoop:` device is `dsnoop:CARD=H1,DEV=0`

"H1" is NOT the name of the sound card -- please stop trying to use it. I showed you that your device name is `dsnoop:CARD=H1,DEV=0` -- if you use A DEVICE NAME, then YES it will work

**Comment by jflyons** [comment]
This is a helpful thread for me...trying to get microphone tuned up on my backyard setup. Thanks!

**Comment by w0mbatian** [comment]
Analysis stops when I set H1 as the recording device in advanced settings.
It starts again if I reset it to default.

In alsamixer the device is detected properly as H1 but if I go out if
alsamixer and then back in again it does not have H1 set and has returned
to saying default.
So there are two things I am asking about.
Thanks
Ian

**Reply by mcguirepr89** [reply]
https://github.com/mcguirepr89/BirdNET-Pi/discussions/551#discussioncomment-3689455

I'm NOT LYING TO YOU!!!!! PLEASE JUST FOLLOW INSTRUCTIONS

**Comment by FirstTenor** [comment]
I think I have a similar problem here.

I installed Birdnet pi last week - using a Pi 4B+, a USB to TRS sound card and a stereo electret mic (water proofed using a yogurt pot and cling film).  Good results all week - despite noticeable mains hum.  We live by the sea - herring gulls are a good loud sound source!

Yesterday I obtained / fitted a ground loop filter in the Audio line - made no difference to monitored sound. … shut system down overnight due to planned power cut.

This morning - no detections after restart.  Alsamixer had switched back to pulse default and sound card seemed to have moved to a different position.  I selected and checked it - but it now just gives me clicks and the hum.

Applied all birdnet updates and restarted.  Again alsamixer reset to default - from this thread, I think that it just does that but all that I now hear is mains hum.  It makes no difference if filter is in or out now.  It also makes no difference how much gain I set.

Need confirmation that whatever usb device is in will select (how do I tell audio is coming in)
Need advice on how to deal with that hum

Thanks

**Reply by mcguirepr89** [reply]
Hi, @FirstTenor -- Have you set the sound card setting in "Tools" > "Settings" > "Advanced Settings" to the device name found when issuing `arecord -L`?

**Reply by FirstTenor** [reply]
Hi @mcguirepr89, Thanks for the tip … Yes, I ran arecord -L and extracted dsnoop:CARD=Device,Dev=0 - this references the USB Audio stream.  I can use alsamixer -m to set the gain on that device - which is up in the ‘white zone’.  Now when I try to listen to ‘live audio’, I get silence.  I still get no detections and the view log for Birdnet-Pi shows a blank screen.

I have also tried swapping microphones - which makes no difference

**Reply by mcguirepr89** [reply]
@FirstTenor have you restarted the core services after setting the sound device? "Tools" > "Services" > "Restart Core Services"?

**Reply by FirstTenor** [reply]
Yes - I tried that first.  I still could hear nothing through the live audio feed (despite the Herring gulls on the roof).  I then stopped and rebooted the whole system.  Same result - and still nothing in the logs either.   Is there some external path that I can check those logs outside of birdnet-pi?

**Reply by mcguirepr89** [reply]
All logging takes place in the system journal, so you can watch it: `journalctl -f`

**Reply by FirstTenor** [reply]
Very useful - thanks …. That showed that there was an error being reported as the number of channels for Audio was wrong - should be one channel even though its a stereo audio adaptor and microphone.  I am now getting a response from the logs - but still no audio monitoring.  Will follow up tomorrow

**Reply by FirstTenor** [reply]
I checked again this morning now that the birds are awake - setting the audio device and correct channel count has fixed the problem.  Many thanks for helpful advice
